### PR TITLE
fix(source): resolve avro `Ref` during `avro_schema_to_column_descs` without hack

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Part of #17020. `Ref` in avro used to be supported by a hack we added in our apache_avro fork. This PR avoids the hack (replacing `Ref` with their referee in-place, resulting in an invalid tree containing duplicate definitions). The hack fails to work when there is `Ref` inside `Ref` - resulting in either unresolved `Ref` or infinite recursion.

This PR only corrects part of the usage - in `avro_schema_to_column_descs` to derive RisingWave column data types from avro schema. There will be a follow-up to correct the usage in `convert_to_datum`/`AvroAccess`. Without the latter part, simple data types like `int` (inside a `Ref` in another `Ref`) can already be supported.

Instead of building the expanded-yet-invalid tree as in the hack, this solution passes a `NamesRef` obtained from the root. Its complexity of having an associated lifetime is easy to deal with in this context. (To contrast, `prost_reflect` for protobuf builds the tree with `Arc` - no lifetime and no duplication. But `apache_avro::Schema` does not use `Arc`.)

The circular reference rejection logic is same as #10499 for protobuf. It can be DRY'ed with error message improved later.

This is intended to be part of v2.2 and NOT cherry-picked into earlier versions. There has been a further hack available for earlier versions that can be cherry-picked on demand - where the user is responsible for not using circular reference or an infinite recursion would happen.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: e2e_test/source_inline/kafka/avro/ref.slt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+
+statement error
+create source s WITH (${RISEDEV_KAFKA_WITH_OPTIONS_COMMON}, topic = 'avro-ref') FORMAT PLAIN ENCODE AVRO (schema.registry = '${RISEDEV_SCHEMA_REGISTRY_URL}');
+----
+db error: ERROR: Failed to run the query
+
+Caused by these errors (recent errors listed first):
+  1: connector error
+  2: circular reference detected: Node -> Node
```
Comment (xxchan) [inline e2e_test/source_inline/kafka/avro/ref.slt:34]: We might need to add to err msg that it's from Avro

**Review by chenzl25 — APPROVED** [review]
rubber stamp

**Review by xxchan — APPROVED** [review]
LGTM

**Comment by xiangjinwu** [comment]
For context, this is the error reported for the added test case if this PR was not applied:
```
ERROR:  Failed to run the query

Caused by these errors (recent errors listed first):
  1: connector error
  2: Feature is not yet implemented: Avro type: Ref { name: Name { name: "Point", namespace: None } }
No tracking issue yet. Feel free to submit a feature request at https://github.com/risingwavelabs/risingwave/issues/new?labels=type%2Ffeature&template=feature_request.yml
```

> There will be a follow-up to correct the usage in `convert_to_datum`/`AvroAccess`. Without the latter part, simple data types like `int` (inside a `Ref` in another `Ref`) can already be supported.

Both parts will be included in v2.2. In this intermediate state, `decimal` inside a `Ref` in another `Ref` would result in a parsing error and filled with null, instead of creation rejected without this first part.

**Pull request merged**